api/mqs/mqs_test.go

```diff
@@ -0,0 +1,630 @@
+		} else if task != nil {
+			t.Errorf("expected nil task but got: %v", task)
+		}
+		if t.Failed() {
+			t.Log(logBuf.String())
+		}
+
+		resetLogBuf()
+
+		// Push invalid (>2) Priority
+		invalidPriority := *task
+		invalidPriority.Priority = new(int32)
+		*invalidPriority.Priority = 5
+		if task, err := mq.Push(ctx, &invalidPriority); err == nil {
+			t.Error("expected invalid priority error but got none")
+		} else if !models.IsErrMQInvalidTaskPriority(err) {
+			t.Errorf("expected invalid priority error but got `%v`", err)
+		} else if task != nil {
+			t.Errorf("expected nil task but got: %v", task)
+		}
+
+		if t.Failed() {
+			t.Log(logBuf.String())
+		}
+	})
+
+	t.Run("delete-invalid", func(t *testing.T) {
+		mq := fixture.newMQ(t)
+
+		// Delete nil Task
+		if err := mq.Delete(ctx, nil); err == nil {
+			t.Errorf("expected error `%v` but got none", models.ErrMQMissingTask)
+		} else if err != models.ErrMQMissingTask {
+			t.Errorf("expected error `%v` but got `%v`", models.ErrMQMissingTask, err)
+		}
+		if t.Failed() {
+			t.Log(logBuf.String())
+		}
+
+		resetLogBuf()
+
+		// Delete empty task
+		if err := mq.Delete(ctx, &models.Task{}); err == nil {
+			t.Errorf("expected error `%v` but got none", models.ErrMQEmptyTaskID)
+		} else if err != models.ErrMQEmptyTaskID {
+			t.Errorf("expected error `%v` but got `%v`", models.ErrMQEmptyTaskID, err)
+		}
+		if t.Failed() {
+			t.Log(logBuf.String())
+		}
+
+		resetLogBuf()
+
+		//Delete non-existent
+		notReal := &models.Task{}
+		notReal.ID = "notReal"
+		notReal.Priority = new(int32)
+		if err := mq.Delete(ctx, notReal); err == nil {
+			t.Errorf("expected error `%v` but got none", models.ErrMQTaskNotReserved)
+		} else if err != models.ErrMQTaskNotReserved {
+			t.Errorf("expected error `%v` but got `%v`", models.ErrMQTaskNotReserved, err)
+		}
+		if t.Failed() {
+			t.Log(logBuf.String())
+		}
+	})
+
+	t.Run("reserve-invalid", func(t *testing.T) {
+		mq := fixture.newMQ(t)
+
+		// Reserve when empty
+		if task, err := mq.Reserve(ctx); err != nil {
+			t.Log(logBuf.String())
+			t.Fatal("unexpected error:", err)
+		} else if task != nil {
+			t.Log(logBuf.String())
+			t.Fatalf("expected nil task but got: %v", task)
+		}
+	})
+
+	t.Run("single-timeout", func(t *testing.T) {
+		mq := fixture.newMQ(t)
+
+		task := &models.Task{}
+		task.ID = "test"
+		task.Priority = new(int32)
+
+		if got, err := mq.Push(ctx, task); err != nil {
+			t.Log(logBuf.String())
+			t.Fatal("unexpected error inserting task:", err)
+		} else if !reflect.DeepEqual(*task, *got) {
+			t.Log(logBuf.String())
+			t.Fatalf("pushed %v but returned %v", task, got)
+		}
+
+		if got, err := mq.Reserve(ctx); err != nil {
+			t.Log(logBuf.String())
+			t.Fatalf("unexpected error on reserve: `%v`", err)
+		} else if !reflect.DeepEqual(*task, *got) {
+			t.Log(logBuf.String())
+			t.Fatalf("pushed %v but reserved %v", task, got)
+		}
+
+		// Let reservation timeout
+		<-time.After(20 * testReserveTimeout)
```

Comment:
It occurred to me that the impls using tickers could be changed to use a manually triggered tick channel for tests, so we can force the background goroutines, rather than try to wait 'long enough'.